yarn-project/archiver/src/archiver/archiver.ts

```diff
@@ -268,46 +261,84 @@ export class Archiver implements ArchiveSource {
+
+    const updateProvenBlock = async () => {
+      // Only update the proven block number if we are behind. And only if we have the state
+      if (provenBlockNumber > BigInt(await this.getProvenBlockNumber())) {
+        const localBlockForDestinationProvenBlockNumber = await this.getBlock(Number(provenBlockNumber));
+        if (
+          localBlockForDestinationProvenBlockNumber &&
+          provenArchive === localBlockForDestinationProvenBlockNumber.archive.root.toString()
+        ) {
+          this.log.info(`Updating the proven block number to ${provenBlockNumber}`);
+          await this.store.setProvenL2BlockNumber(Number(provenBlockNumber));
+        }
+      }
+    };
 
-    if (noBlocksButInitial || noBlockSinceLast) {
+    // This is an edge case that we only hit if there are no proposed blocks.
+    const noBlocks = localPendingBlockNumber === 0n && pendingBlockNumber === 0n;
+    if (noBlocks) {
       await this.store.setBlockSynchedL1BlockNumber(currentL1BlockNumber);
       this.log.verbose(`No blocks to retrieve from ${blocksSynchedTo + 1n} to ${currentL1BlockNumber}`);
       return;
     }
 
-    if (lastBlock && archiveOfMyBlock !== lastBlock.archive.root.toString()) {
-      // @todo  Either `prune` have been called, or L1 have re-orged deep enough to remove a block.
-      //        Issue#8620 and Issue#8621
+    // Related to the L2 reorgs of the pending chain. We are only interested in actually addressing a reorg if there
+    // are any state that could be impacted by it. If we have no blocks, there is no impact.
+    if (localPendingBlockNumber > 0) {
+      const localPendingBlock = await this.getBlock(Number(localPendingBlockNumber));
+      if (localPendingBlock === undefined) {
+        throw new Error(`Missing block ${localPendingBlockNumber}`);
+      }
+
+      const noBlockSinceLast = localPendingBlock && pendingArchive === localPendingBlock.archive.root.toString();
+      if (noBlockSinceLast) {
+        // While there have been no L2 blocks, there might have been a proof, so we will update if needed.
+        await updateProvenBlock();
+        await this.store.setBlockSynchedL1BlockNumber(currentL1BlockNumber);
+        this.log.verbose(`No blocks to retrieve from ${blocksSynchedTo + 1n} to ${currentL1BlockNumber}`);
+        return;
+      }
+
+      const localPendingBlockInChain = archiveForLocalPendingBlockNumber === localPendingBlock.archive.root.toString();
+      if (!localPendingBlockInChain) {
+        // If our local pending block tip is not in the chain on L1 a "prune" must have happened
+        // or the L1 have reorged.
+        // In any case, we have to figure out how far into the past the action will take us.
+        // For simplicity here, we will simply rewind until we end in a block that is also on the chain on L1.
+        this.log.verbose(`L2 prune have occurred, unwind state`);
+
+        let tipAfterUnwind = localPendingBlockNumber;
+        while (true) {
```

Comment:
This condition can be hit if the Pending chain reorgs from a prune

In which case the entire epoch will be dropped, correct me if im wrong but if this is the median case then we would make less requests if we start from the proven chain then move forward?

Reply:
Essentially yes. I think I had a weird edge case where I needed to figure out whether to use the local proven or the destination proven and were like, I will just take the simple thing that works and then deal with optimisations later if they are really needed. 

If running the `updateProvenBlock` in the case of `localPendingBlockInChain` that should probably be covered though. 

Separately, I think when I was doing it, part of having it like this was the case where the l2 prune was caused by an L1 reorg such that the proven local chain was invalid and would also end up being unwound here, but that would not actually update our local proven block number in the same round which is kinda odd 🤷. Actually the `updateProvenBlock` should probably not only update if provenBlock have moved ahead since that is going to kinda nuke itself with a reorg caused by L1.